Library/Homebrew/cmd/autoremove.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+require "formula"
+require "cli/parser"
+
+module Homebrew
+  module_function
+
+  def autoremove_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `autoremove` [<options>]
+
+        Remove packages that weren't installed on request and are no longer needed.
+      EOS
+      switch "-n", "--dry-run",
+             description: "Just print what would be removed."
+      switch "-f", "--force",
+             description: "Remove without confirmation."
+      named 0
+    end
+  end
+
+  def get_removable_formulae(installed_formulae)
```

Comment:
This function seems to intersect with what `leaves.rb` does. Can/should we factor out common behavior to use the same logic in both places?

Reply:
There is not that much shared logic between those two files, I think.

Reply:
I think it would be good to share the logic that is there e.g.
```ruby
    installed = Formula.installed.sort
    deps_of_installed = installed.flat_map(&:runtime_formula_dependencies)
    leaves = installed.map(&:full_name) - deps_of_installed.map(&:full_name)
```

Reply:
And also be able to write tests for that shared logic.

Reply:
I don't think the logic between `leaves.rb` and `autoremove.rb` can be shared. `leaves.rb` only lists out the top level of formulae that are not dependencies of another formula. But `autoremove.rb` is recursive and also lists/removes dependencies that are no longer needed after dependant is removed.

Reply:
I think some logic can still be shared because `leaves` will/should never be `autoremove`d.

Reply:
I was actually thinking the opposite. Only `leaves` that were not *installed_on_request* can be `autoremove`d. Because otherwise, it would be a dependency and should never be `autoremove`d until the dependent is removed.